File: script/clear_trans_html.py
#!/usr/bin/env python3
"""clear_trans_html.py <url> —— Python 工作流：完整性→特殊元素→清理→readability-lxml→markdownify→sketch.md"""
import json
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent))

from markdownify import MarkdownConverter  # noqa: E402
from readability import Document  # noqa: E402

from pylib import env, placeholder  # noqa: E402
from pylib.browser import open_page  # noqa: E402


def log_err(*a):
    print(*a, file=sys.stderr)


def emit(result: dict, code: int):
    print(json.dumps(result, ensure_ascii=False), flush=True)
    sys.exit(code)


def usage(msg: str):
    emit({"status": "usage_error", "reason": msg}, 2)


def progressive_scroll(page):
    page.evaluate("""async () => {
    let last = -1;
    for (let i = 0; i < 60; i++) {
      window.scrollTo(0, document.body.scrollHeight);
      await new Promise((r) => setTimeout(r, 150));
      const h = document.documentElement.scrollHeight;
      if (h === last) break;
      last = h;
    }
    window.scrollTo(0, 0);
  }""")


def wait_for_dom_stable(page, stable_ms=1000, max_ms=15000):
    import time
    # Node 版用 Date.now()（毫秒）；time.time() 为秒，阈值须换算，否则稳定页要自旋 1000 秒
    stable_s, max_s = stable_ms / 1000, max_ms / 1000
    t0 = time.time()
    last, last_change = -1, time.time()
    while time.time() - t0 < max_s:
        n = page.evaluate("() => document.getElementsByTagName('*').length")
        if n != last:
            last, last_change = n, time.time()
        elif time.time() - last_change >= stable_s:
            return
        page.wait_for_timeout(200)


def code_language(tag):
    """markdownify 回调：从 <code class="language-x"> 补齐围栏语言标注（与 Node 工作流形态一致）。

    回调收到的是 <pre>（块级）本身，语言标注在其子 <code> 上，故两者都查。
    """
    for el in (tag, *tag.find_all("code")):
        for cls in el.get("class") or []:
            if cls.startswith("language-"):
                return cls[len("language-"):]
    return ""


def main():
    if len(sys.argv) < 2 or sys.argv[1].startswith("--"):
        usage("用法: clear_trans_html.py <url>")
    url = sys.argv[1]
    dirs = env.ensure_workflow_dirs(url, "python_workflow")
    warnings = []

    page_init = placeholder.read_shared_script("page-init.js")
    page_merge = placeholder.read_shared_script("page-merge.js")
    page_clean = placeholder.read_shared_script("page-clean.js")

    session = open_page(url, viewport={"width": 1280, "height": 3000},
                        init_scripts=[page_init],
                        storage_state_path=env.storage_state_path(), log=log_err)
    try:
        page = session.page
        progressive_scroll(page)
        wait_for_dom_stable(page)

        ctx = placeholder.make_ctx(dirs, session.context, log_err)
        page.evaluate(f"({page_merge})()")
        placeholder.process_mermaid(page, ctx)
        placeholder.process_special_elements(page, ctx)
        placeholder.process_images(page, ctx)
        page.evaluate(f"({page_clean})()")

        # 不传 url：readability-lxml 仅在给定 url 时才 make_links_absolute，
        # 分派自产的 assets/ 相对引用须原样保留（Node 侧 Readability 的绝对化在 mjs 中另行还原）
        # min_text_length=10：默认 25 会把 sanitize 中 content_length<25 的 table/div
        # （表格、"graph TD" 级短代码块容器）当噪声剔除，@mozilla/readability 无此行为
        html = page.content()
        doc = Document(html, min_text_length=10)
        content = doc.summary(html_partial=True)
        if not content or len(content.strip()) < 20:
            warnings.append("readability-lxml 未能解析主体，回退 body 全文")
            content = page.evaluate("() => document.body.innerHTML")

        md = MarkdownConverter(heading_style="ATX", bullets="-",
                               code_language_callback=code_language).convert(content)
        # 不转义下划线（镜像 Node .replace(/\\_/g, '_')）：{{IMG_n}}/{{COMPLEX_DIV_n}}
        # 是后续精确替换的机器令牌，markdownify 默认转义 _ 须还原
        md = md.replace("\\_", "_").strip() + "\n"

        (dirs["wf"] / "sketch.md").write_text(md, encoding="utf-8")
        placeholder.write_manifest(dirs["manifest"], ctx["entries"])
        payload = {
            "status": "ok",
            "sketch": str(dirs["wf"] / "sketch.md"),
            "images": ctx["counters"]["img"],
            "complex": len(ctx["entries"]),
            "warnings": warnings + ctx["warnings"],
        }
    finally:
        try:
            session.close()  # 先关浏览器再输出退出（emit 内 sys.exit）；关闭失败不吞结果（镜像 Node .catch(() => {})）
        except Exception:  # noqa: BLE001
            pass
    emit(payload, 0)


if __name__ == "__main__":
    try:
        main()
    except SystemExit:
        raise
    except Exception as e:  # noqa: BLE001
        emit({"status": "error", "reason": str(e)}, 1)

File: script/clear_trans_html.mjs
#!/usr/bin/env node
// clear_trans_html.mjs <url> —— Node 工作流：完整性→特殊元素→清理→readability→turndown→sketch.md
import path from 'node:path';
import { emit, emitError, usage, log } from './lib/contract.mjs';
import { projectRoot, ensureWorkflowDirs, storageStatePath } from './lib/env.mjs';
import { openPage } from './lib/browser.mjs';
import { makeCtx, readSharedScript, processMermaid, processSpecialElements, processImages, writeManifest } from './lib/placeholder.mjs';
import TurndownService from 'turndown';
import { gfm } from '@joplin/turndown-plugin-gfm';
import fs from 'node:fs/promises';

const READABILITY_JS = path.join(projectRoot(), 'node_modules', '@mozilla', 'readability', 'Readability.js');

/** 渐进滚动到底再回顶（懒加载） */
async function progressiveScroll(page) {
  await page.evaluate(async () => {
    let last = -1;
    for (let i = 0; i < 60; i++) {
      window.scrollTo(0, document.body.scrollHeight);
      await new Promise((r) => setTimeout(r, 150));
      const h = document.documentElement.scrollHeight;
      if (h === last) break;
      last = h;
    }
    window.scrollTo(0, 0);
  });
}

/** DOM 稳定：节点数连续 stableMs 不变（虚拟 DOM 场景；不移除任何元素） */
async function waitForDomStable(page, { stableMs = 1000, maxMs = 15000 } = {}) {
  const t0 = Date.now();
  let last = -1;
  let lastChange = Date.now();
  while (Date.now() - t0 < maxMs) {
    const n = await page.evaluate(() => document.getElementsByTagName('*').length);
    if (n !== last) { last = n; lastChange = Date.now(); }
    else if (Date.now() - lastChange >= stableMs) return;
    await page.waitForTimeout(200);
  }
}

async function main() {
  const url = process.argv[2];
  if (!url || url.startsWith('--')) return usage('用法: clear_trans_html.mjs <url>');
  const dirs = ensureWorkflowDirs(url, 'node_workflow');

  const pageInit = await readSharedScript('page-init.js');
  const pageMerge = await readSharedScript('page-merge.js');
  const pageClean = await readSharedScript('page-clean.js');

  let s;
  let result;
  try {
    s = await openPage(url, { viewport: { width: 1280, height: 3000 }, initScripts: [pageInit], storageStatePath: storageStatePath(), log });
    await progressiveScroll(s.page);
    await waitForDomStable(s.page);

    const ctx = makeCtx(dirs, { context: s.context, log });
    await s.page.evaluate(`(${pageMerge})()`);
    await processMermaid(s.page.mainFrame(), ctx);
    await processSpecialElements(s.page.mainFrame(), ctx);
    await processImages(s.page.mainFrame(), ctx);

    await s.page.evaluate(`(${pageClean})()`);

    // Readability 在页面内运行（避免 jsdom 依赖）
    await s.page.addScriptTag({ path: READABILITY_JS });
    // keepClasses: 默认会剥除全部 class，fenced 代码块语言标注（language-*）将丢失
    const article = await s.page.evaluate(() => {
      const a = new Readability(document, { keepClasses: true }).parse();
      return a ? { title: a.title, content: a.content } : null;
    });
    let html;
    if (article?.content) {
      html = article.content;
    } else {
      ctx.warnings.push('readability 未能解析主体，回退 body 全文');
      html = await s.page.evaluate(() => document.body.innerHTML);
    }
    // Readability _fixRelativeUris 会把分派自产的 assets/ 相对引用绝对化——按 manifest 还原
    for (const e of ctx.entries) {
      if (e.final) html = html.split(new URL(e.final, url).href).join(e.final);
    }

    const td = new TurndownService({ codeBlockStyle: 'fenced', headingStyle: 'atx', bulletListMarker: '-' });
    td.use(gfm);
    // 不转义下划线：{{IMG_n}}/{{COMPLEX_DIV_n}} 是后续精确替换的机器令牌，须原样保留
    const md = td.turndown(html).replace(/\\_/g, '_');

    await fs.writeFile(path.join(dirs.wf, 'sketch.md'), md, 'utf8');
    await writeManifest(dirs.manifest, ctx.entries);
    result = {
      status: 'ok',
      sketch: path.join(dirs.wf, 'sketch.md'),
      images: ctx.counters.img,
      complex: ctx.entries.length,
      warnings: ctx.warnings,
    };
  } catch (e) {
    await s?.close().catch(() => {});
    return emitError(e.message, 1);
  }
  await s.close().catch(() => {}); // 先关浏览器再 emit（emit 内 process.exit）
  emit(result, 0);
}

main().catch((e) => emitError(e.message, 1));
